src/lua/protobuf_wireformat.lua

```diff
@@ -152,6 +283,28 @@ local function encode_fixed32(field_id, value)
     return encode_tag(field_id, WIRE_TYPE_I32) .. as_int32(value)
 end
 
+-- Decode an unsigned integral value encoded as I32.
+--
+-- Input value type: string.
+--
+-- Used for Protocol Buffers types: fixed32.
+local function decode_fixed32(ctx)
+    local res = string.sub(ctx.val, ctx.pos, ctx.pos + 3)
+    ctx.pos = ctx.pos + 4
+    return tonumber(from_int32(res))
+end
+
+-- Decode a signed integral value encoded as I32.
+--
+-- Input value type: string.
+--
+-- Used for Protocol Buffers types: sfixed32.
+local function decode_sfixed32(ctx)
+    local res = string.sub(ctx.val, ctx.pos, ctx.pos + 3)
+    ctx.pos = ctx.pos + 4
+    return tonumber(ffi.cast('int32_t', from_int32(res)))
```

Comment:
[n1256.pdf](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf), 6.3.1.3 Signed and unsigned integers (page 55 as shown by my pdf reader, page 43 in on the paper):

> 6.3.1.3 Signed and unsigned integers
>
> 1. When a value with integer type is converted to another integer type other than _Bool, if the value can be represented by the new type, it is unchanged.
> 2. Otherwise, if the new type is unsigned, the value is converted by repeatedly adding or subtracting one more than the maximum value that can be represented in the new type
until the value is in the range of the new type.<sup>49)</sup>
> 3. Otherwise, the new type is signed and the value cannot be represented in it; either the result is implementation-defined or an implementation-defined signal is raised.
>
> 49\) The rules describe arithmetic on the mathematical value, not the value of a given type of expression.

The bullet 2 is not applicable (the new type is signed). If the value is above 2^31-1, the result is implementation defined according to the bullet 3.

I would avoid casting from unsigned type to a signed type with intention to get possibly negative values.

I think that we should have two different functions that convert 4 bytes in a string to a number: for example, `extract_uint32` and `extract_int32`. Both should return `number` (if we're about the 32 bit variants).

Reply:
Added two different functions to extract signed and unsigned integers from decoded string.